core/modules/file/file.module

```diff
@@ -2777,101 +2777,68 @@ function file_get_type($file) {
- *   (e.g. 'image') for "create" operation.
- * @param $account
+ *   (e.g. 'image').
+ * @param object $account
  *   Optional, a user object representing the user for whom the operation is to
  *   be performed. Determines access for a user other than the current user.
  *
- * @return
+ * @return bool
  *   TRUE if the operation may be performed, FALSE otherwise.
  */
 function file_access($op, $file = NULL, $account = NULL) {
-  $rights = &backdrop_static(__FUNCTION__, array());
-
-  if (!$file && !in_array($op, array('view', 'download', 'update', 'delete', 'create'), TRUE)) {
-    // If there was no file to check against, and the $op was not one of the
-    // supported ones, we return access denied.
+  // Filter out unsupported operations.
+  if (!in_array($op, array('create', 'view', 'download', 'update', 'delete'), TRUE)) {
     return FALSE;
   }
 
-  // If no user object is supplied, the access check is for the current user.
-  if (empty($account)) {
-    $account = $GLOBALS['user'];
-  }
-
-  // $file may be either an object or a file type. Since file types cannot be
-  // an integer, use either fid or type as the static cache id.
-  $cache_id = NULL;
-  if (!empty($file->id)) {
-    $cache_id = $file->id;
-  }
-  elseif (is_string($file) && $op == 'create') {
-    $cache_id = $file;
-  }
-  else {
-    $cache_id = backdrop_hash_base64(serialize($file));
-  }
-
-  // If we've already checked access for this file, user and op, return from
-  // cache.
-  if (isset($rights[$account->uid][$cache_id][$op])) {
-    return $rights[$account->uid][$cache_id][$op];
-  }
+  // If no $file is given or it's only the file type.
+  if (!$file || is_string($file)) {
+    // Filter out unsupported file types.
+    if (is_string($file) && !in_array($file, file_permissions_get_configured_types())) {
+      return FALSE;
+    }
 
-  if (user_access('bypass file access', $account)) {
-    return $rights[$account->uid][$cache_id][$op] = TRUE;
-  }
+    // No restrictions for this account.
+    if (user_access('bypass file access', $account)) {
+      return TRUE;
+    }
 
-  // We grant access to the file if both of the following conditions are met:
-  // - No modules say to deny access.
-  // - At least one module says to grant access.
-  $access = module_invoke_all('file_access', $op, $file, $account);
-  if (in_array(FILE_ACCESS_DENY, $access, TRUE)) {
-    return $rights[$account->uid][$cache_id][$op] = FALSE;
-  }
-  elseif (in_array(FILE_ACCESS_ALLOW, $access, TRUE)) {
-    return $rights[$account->uid][$cache_id][$op] = TRUE;
-  }
+    // Permission checks that work without a file instance.
+    switch ($op) {
+      case 'create':
+        return user_access('create files', $account);
 
+      case 'view':
+        // Without instance we can not know if a file might be public or
+        // private, so we require both permissions.
+        return user_access('view files', $account) && user_access('view private files', $account);
```

Comment:
I'm actually not sure if this is sufficient, as the file status (temporary/permanent) also plays a role. And the status isn't available without an instance.